### PR TITLE
terminal/logflags: Added `SetLoggerFactory(LoggerFactory)`

### PR DESCRIPTION
**Pull request by blaubaer**
This change will enable people who want to embed Delve into their applications to adjust the logging better to their needs.

### CODE REVIEW

**Comment by blaubaer** [comment]
Does someone know why the test `TestClientServer_FindLocationsAddr` (in this run) is failing? I cannot see a reason why it should fail (caused by my changes) nor I can bring it to fail locally. Is this a temporary issue?

**Comment by aarzilli** [comment]
> Does someone know why the test `TestClientServer_FindLocationsAddr` (in this run) is failing? I cannot see a reason why it should fail (caused by my changes) nor I can bring it to fail locally. Is this a temporary issue?

The windows/arm64 port has some problems.

File: pkg/logflags/logflags.go [hunk 1]
```diff
@@ -217,12 +236,17 @@ func Close() {
 	}
 }
 
-// textFormatter is a simplified version of logrus.TextFormatter that
+// DefaultFormatter provides a simplified version of logrus.TextFormatter that
 // doesn't make logs unreadable when they are output to a text file or to a
 // terminal that doesn't support colors.
-type textFormatter struct {
+func DefaultFormatter() logrus.Formatter {
```
Comment (aarzilli) [inline pkg/logflags/logflags.go:242]: Why do we want to export this?

File: pkg/logflags/logflags.go [hunk 2]
```diff
@@ -29,17 +29,24 @@ var minidump = false
 
 var logOut io.WriteCloser
 
-func makeLogger(flag bool, fields logrus.Fields) *logrus.Entry {
-	logger := logrus.New().WithFields(fields)
-	logger.Logger.Formatter = &textFormatter{}
+func makeFlaggableLogger(flag bool, fields Fields) Logger {
```
Comment (aarzilli) [inline pkg/logflags/logflags.go:32]: No reason to split this function in to two functions when you only call makeFlaggableLogger anyway.

File: pkg/logflags/logflags.go [hunk 3]
```diff
@@ -80,14 +87,27 @@ func DebugLineErrors() bool {
...
-	return makeLogger(rpc, logrus.Fields{"layer": "rpc"})
+func RPCLogger() Logger {
+	if rpc {
+		return RPCLoggerWithLevel(logrus.DebugLevel)
+	}
+	return RPCLoggerWithLevel(logrus.ErrorLevel)
+}
+
+// RPCLoggerWithLevel returns a logger for RPC messages set to a specific minimal log level.
+func RPCLoggerWithLevel(level logrus.Level) Logger {
```
Comment (aarzilli) [inline pkg/logflags/logflags.go:109]: No reason to export this.

File: pkg/logflags/logflags.go [hunk 4]
```diff
@@ -139,16 +159,15 @@ func writeListeningMessage(server string, addr net.Addr) {
...
-	logger := RPCLogger()
-	logger.Logger.Level = logrus.WarnLevel
+	logger := RPCLoggerWithLevel(logrus.WarnLevel)
 	logger.Warnln("Listening for remote connections (connections are not authenticated nor encrypted)")
 }
 
 func WriteError(msg string) {
 	if logOut != nil {
-		fmt.Fprintln(logOut, msg)
+		_, _ = fmt.Fprintln(logOut, msg)
```
Comment (aarzilli) [inline pkg/logflags/logflags.go:168]: We don't do this `_, _ = fmt.Fprintln` thing anywhere.

File: pkg/logflags/logger.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	Errorln(args ...interface{})
+	Fatalln(args ...interface{})
+	Panicln(args ...interface{})
+}
+
+// LoggerFactory is used to create new logrus.FieldLogger instances.
+// SetLoggerFactory can be used to configure it.
+//
+// The given parameters fields and out can be both be nil.
+type LoggerFactory func(level logrus.Level, fields Fields, out io.Writer) Logger
```
Comment (aarzilli) [inline pkg/logflags/logger.go:50]: This function shouldn't take a logrus.Level, it should take the bool flag that says if logging is enabled or not.

File: pkg/logflags/logflags.go (hunk 3 above)
Comment (blaubaer) [reply pkg/logflags/logflags.go:109]: Ok.

File: pkg/logflags/logflags.go (hunk 4 above)
Comment (blaubaer) [reply pkg/logflags/logflags.go:168]: Ok.

File: pkg/logflags/logflags.go (hunk 2 above)
Comment (blaubaer) [reply pkg/logflags/logflags.go:32]: This is caused by [`pkg/logflags/logflags.go:163`](https://github.com/go-delve/delve/pull/3257/files/ccda6c241cf732d8ce033eb443f31beead8da3d7#diff-f89e32a213ab96ab1585632446e21aef2976688c4ac06c461d6ceb95bcc2903cR163):
```golang
logger := RPCLoggerWithLevel(logrus.WarnLevel)
logger.Warnln("Listening for remote connections (connections are not authenticated nor encrypted)")
```

In this case the initial code forced that this Warning should always be displayed, although if tagged or not. With this I tried to preserve this behavior.

Comment (aarzilli) [reply pkg/logflags/logflags.go:32]: We can just pass true as the flag.

Comment (blaubaer) [reply pkg/logflags/logflags.go]: Ok.

File: pkg/logflags/logger.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	Infoln(args ...interface{})
+	Println(args ...interface{})
+	Warnln(args ...interface{})
+	Warningln(args ...interface{})
+	Errorln(args ...interface{})
+	Fatalln(args ...interface{})
+	Panicln(args ...interface{})
+}
+
+// LoggerFactory is used to create new logrus.FieldLogger instances.
```
Comment (aarzilli) [inline pkg/logflags/logger.go:46]: s/new logrus.FieldLogger instances/new Logger instances/

**Review by aarzilli — APPROVED** [review]
LGTM

**Review by derekparker — APPROVED** [review]
LGTM

**Pull request merged**